### PR TITLE
Fix circular dependency and refactor Log

### PR DESCRIPTION
**Pull request by LaurenNguyen14**
Apart from fixing the circular dependency, the goal of refactoring C4Log and Log is also to make all getters/setters implementation of C4Log stay within the C4 class level

### CODE REVIEW

File: common/main/java/com/couchbase/lite/FileLogger.java [hunk 1]
```diff
@@ -57,7 +57,7 @@
     @Override
     public void log(@NonNull LogLevel level, @NonNull LogDomain domain, @NonNull String message) {
         if ((config == null) || (level.compareTo(logLevel) < 0)) { return; }
-        c4Log.logToCore(Log.getC4DomainForLoggingDomain(domain), Log.getC4LevelForLogLevel(level), message);
+        c4Log.logToCore(c4Log.getC4DomainForLoggingDomain(domain), c4Log.getC4LevelForLogLevel(level), message);
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/FileLogger.java:60]: I think this should just be:
c4Log.logToCore(domain, level, message)

File: common/main/java/com/couchbase/lite/FileLogger.java [hunk 2]
```diff
@@ -77,7 +77,7 @@ public void setLevel(@NonNull LogLevel level) {
         if (logLevel == level) { return; }
         logLevel = level;
 
-        if (!initLog()) { c4Log.setFileFileLevel(Log.getC4LevelForLogLevel(level)); }
+        if (!initLog()) { c4Log.setFileFileLevel(c4Log.getC4LevelForLogLevel(level)); }
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/FileLogger.java:80]: again: c4Log.setFileLevel(level)

File: common/main/java/com/couchbase/lite/FileLogger.java [hunk 3]
```diff
@@ -146,7 +146,7 @@ private boolean initLog() {
 
         c4Log.initFileLogger(
             logDirPath,
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/FileLogger.java:148]: This one too.  Call it with the LogLevel and let c4Log do the conversion

File: common/main/java/com/couchbase/lite/internal/core/C4Log.java [hunk 4]
```diff
@@ -19,6 +19,10 @@
 import androidx.annotation.Nullable;
 import androidx.annotation.VisibleForTesting;
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4Log.java:21]: NICE!!!

File: common/main/java/com/couchbase/lite/internal/core/C4Log.java [hunk 5]
```diff
@@ -86,10 +142,82 @@ public final void forceCallbackLevel(@NonNull LogLevel logLevel) {
         setCoreCallbackLevel();
     }
 
+    @VisibleForTesting
+    // This, apparently, should be the inverse of LOGGING_DOMAINS_FROM_C4
+    public final void setC4LogLevel(@NonNull EnumSet<LogDomain> domains, @NonNull LogLevel level) {
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4Log.java:147]: I think this doesn't need to be public, anymore

File: common/main/java/com/couchbase/lite/internal/core/C4Log.java [hunk 6]
```diff
@@ -86,10 +142,82 @@ public final void forceCallbackLevel(@NonNull LogLevel logLevel) {
...
+                default:
+                    logInternal(
+                        getC4DomainForLoggingDomain(LogDomain.DATABASE),
+                        c4Level,
+                        "Unexpected log domain: " + domain);
+                    break;
+            }
+        }
+    }
+
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4Log.java:193]: The rest of these, too.  I'll bet they don't need to be public anymore...

File: common/main/java/com/couchbase/lite/internal/support/Log.java [hunk 7]
```diff
@@ -135,7 +77,7 @@ public static void initLogging(@NonNull Map<String, String> errorMessages) {
     @VisibleForTesting
     public static void initLogging() {
         C4Log.get().forceCallbackLevel(Database.log.getConsole().getLevel());
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/support/Log.java:79]: I think you probably want to get it once and assign it to a local variable...

**Review by bmeike — COMMENTED** [review]
It looks great.  I think you just didn't take the refactor far enough.  There is, really, never any reason to do:
foo(foo.this(x), foo.that(y)).
It can *always* be refactored to:
foo(x, y)

**Review by bmeike — CHANGES_REQUESTED** [review]
Request that you finish the refactor.

File: common/main/java/com/couchbase/lite/internal/core/C4Log.java (hunk 5 above)
Comment (LaurenNguyen14) [reply common/main/java/com/couchbase/lite/internal/core/C4Log.java:147]: support.Log calls this method in initLogging()

`    @VisibleForTesting
    public static void initLogging() {
        C4Log.get().forceCallbackLevel(Database.log.getConsole().getLevel());
        C4Log.get().setC4LogLevel(LogDomain.ALL_DOMAINS, LogLevel.DEBUG);
    }
`

Comment (LaurenNguyen14) [reply common/main/java/com/couchbase/lite/internal/core/C4Log.java:147]: support.Log calls this method in initLogging()

`    @VisibleForTesting
    public static void initLogging() {
        C4Log.get().forceCallbackLevel(Database.log.getConsole().getLevel());
        C4Log.get().setC4LogLevel(LogDomain.ALL_DOMAINS, LogLevel.DEBUG);
    }
`

File: common/main/java/com/couchbase/lite/internal/core/C4Log.java (hunk 6 above)
Comment (LaurenNguyen14) [reply common/main/java/com/couchbase/lite/internal/core/C4Log.java:195]: one/two of these methods are used by LogTest also

File: common/main/java/com/couchbase/lite/internal/core/C4Log.java (hunk 5 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/internal/core/C4Log.java:149]: Then it isn't visible for testing, is it?

File: common/main/java/com/couchbase/lite/internal/core/C4Log.java [hunk 8]
```diff
@@ -46,21 +49,76 @@ public static void logCallback(@NonNull String c4Domain, int c4Level, @NonNull S
...
+        LOG_LEVEL_TO_C4 = Collections.unmodifiableMap(m);
+    }
+    public final void logToCore(LogDomain domain, LogLevel level, String message) {
+        log(getC4DomainForLoggingDomain(domain), getC4LevelForLogLevel(level), message);
+    }
 
     public final int getFileLogLevel() { return getBinaryFileLevel(); }
 
-    public final void setFileFileLevel(int level) { setBinaryFileLevel(level); }
+    public final void setFileFileLevel(LogLevel level) { setBinaryFileLevel(getC4LevelForLogLevel(level)); }
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4Log.java:112]: formatting

File: common/main/java/com/couchbase/lite/FileLogger.java [hunk 9]
```diff
@@ -57,7 +57,8 @@
     @Override
     public void log(@NonNull LogLevel level, @NonNull LogDomain domain, @NonNull String message) {
         if ((config == null) || (level.compareTo(logLevel) < 0)) { return; }
-        c4Log.logToCore(Log.getC4DomainForLoggingDomain(domain), Log.getC4LevelForLogLevel(level), message);
+        //c4Log.logToCore(c4Log.getC4DomainForLoggingDomain(domain), c4Log.getC4LevelForLogLevel(level), message);
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/FileLogger.java:60]: Are you trying to troll me?  You can't submit commented out code...

**Review by bmeike — APPROVED** [review]
Fix the silly mistakes and commit it.

**Pull request merged**